File: DebtCompass/DebtCompass/Engine/PayoffStrategy.swift
import Foundation

enum PayoffStrategy: String, CaseIterable, Identifiable {
    case avalanche
    case snowball
    case velocityBanking

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .avalanche: return "Debt Avalanche"
        case .snowball: return "Debt Snowball"
        case .velocityBanking: return "Velocity Banking"
        }
    }

    var summary: String {
        switch self {
        case .avalanche:
            return "Pays extra toward the highest interest-rate debt first. Mathematically minimizes total interest paid."
        case .snowball:
            return "Pays extra toward the smallest balance first. Builds momentum with quick wins, at the cost of a little extra interest."
        case .velocityBanking:
            return "Uses a line of credit to knock out one debt at a time in large chunks, then aggressively pays down the line of credit with all monthly cash flow before moving to the next debt."
        }
    }
}

/// Immutable snapshot of a debt at a point in the simulation.
struct DebtSnapshot: Identifiable {
    let id: UUID
    let name: String
    let type: DebtType
    var balance: Double
    let apr: Double
    let minimumPayment: Double
}

/// One simulated month of a payoff plan.
struct PayoffMonthEntry: Identifiable {
    var id: Int { month }
    let month: Int
    /// Interest accrued this month, per debt id.
    let interestByDebt: [UUID: Double]
    /// Total paid toward each debt this month (minimum + any extra/velocity payment).
    let paymentByDebt: [UUID: Double]
    /// Remaining balance at the end of the month, per debt id.
    let remainingBalanceByDebt: [UUID: Double]

    var totalInterest: Double { interestByDebt.values.reduce(0, +) }
    var totalPayment: Double { paymentByDebt.values.reduce(0, +) }
    var totalRemainingBalance: Double { remainingBalanceByDebt.values.reduce(0, +) }
}

struct PayoffPlanResult {
    let strategy: PayoffStrategy
    let debtNames: [UUID: String]
    let months: [PayoffMonthEntry]
    let totalInterestPaid: Double
    let totalPaid: Double
    /// True if minimum payments don't cover accruing interest and the simulation hit its
    /// safety cap without reaching zero balance.
    let neverPaysOff: Bool

    var monthsToPayoff: Int { months.count }

    var yearsAndMonthsToPayoff: (years: Int, months: Int) {
        (monthsToPayoff / 12, monthsToPayoff % 12)
    }

    /// The month each debt reaches a zero balance, keyed by debt id.
    var payoffMonthByDebt: [UUID: Int] {
        var result: [UUID: Int] = [:]
        for entry in months {
            for (debtID, balance) in entry.remainingBalanceByDebt where balance <= 0.005 {
                if result[debtID] == nil { result[debtID] = entry.month }
            }
        }
        return result
    }
}

File: web/src/engine/payoffStrategy.ts
import type { DebtType } from "../models/types";

export type PayoffStrategy = "avalanche" | "snowball" | "velocityBanking";

export const PAYOFF_STRATEGIES: PayoffStrategy[] = ["avalanche", "snowball", "velocityBanking"];

export const PAYOFF_STRATEGY_LABELS: Record<PayoffStrategy, string> = {
  avalanche: "Debt Avalanche",
  snowball: "Debt Snowball",
  velocityBanking: "Velocity Banking",
};

export const PAYOFF_STRATEGY_SUMMARIES: Record<PayoffStrategy, string> = {
  avalanche:
    "Pays extra toward the highest interest-rate debt first. Mathematically minimizes total interest paid.",
  snowball:
    "Pays extra toward the smallest balance first. Builds momentum with quick wins, at the cost of a little extra interest.",
  velocityBanking:
    "Uses a line of credit to knock out one debt at a time in large chunks, then aggressively pays down the line of credit with all monthly cash flow before moving to the next debt.",
};

/** Immutable snapshot of a debt at a point in the simulation. */
export interface DebtSnapshot {
  id: string;
  name: string;
  type: DebtType;
  balance: number;
  apr: number;
  minimumPayment: number;
}

/** One simulated month of a payoff plan. */
export interface PayoffMonthEntry {
  month: number;
  /** Interest accrued this month, per debt id. */
  interestByDebt: Record<string, number>;
  /** Total paid toward each debt this month (minimum + any extra/velocity payment). */
  paymentByDebt: Record<string, number>;
  /** Remaining balance at the end of the month, per debt id. */
  remainingBalanceByDebt: Record<string, number>;
}

export function monthTotalInterest(m: PayoffMonthEntry): number {
  return Object.values(m.interestByDebt).reduce((a, b) => a + b, 0);
}

export function monthTotalPayment(m: PayoffMonthEntry): number {
  return Object.values(m.paymentByDebt).reduce((a, b) => a + b, 0);
}

export function monthTotalRemainingBalance(m: PayoffMonthEntry): number {
  return Object.values(m.remainingBalanceByDebt).reduce((a, b) => a + b, 0);
}

export interface PayoffPlanResult {
  strategy: PayoffStrategy;
  debtNames: Record<string, string>;
  months: PayoffMonthEntry[];
  totalInterestPaid: number;
  totalPaid: number;
  /** True if minimum payments don't cover accruing interest and the simulation
   * hit its safety cap without reaching zero balance. */
  neverPaysOff: boolean;
}

export function monthsToPayoff(result: PayoffPlanResult): number {
  return result.months.length;
}

export function yearsAndMonthsToPayoff(result: PayoffPlanResult): { years: number; months: number } {
  const total = monthsToPayoff(result);
  return { years: Math.floor(total / 12), months: total % 12 };
}

/** The month each debt reaches a zero balance, keyed by debt id. */
export function payoffMonthByDebt(result: PayoffPlanResult): Record<string, number> {
  const out: Record<string, number> = {};
  for (const entry of result.months) {
    for (const [debtId, balance] of Object.entries(entry.remainingBalanceByDebt)) {
      if (balance <= 0.005 && out[debtId] === undefined) {
        out[debtId] = entry.month;
      }
    }
  }
  return out;
}
